File: ScheduleGenerator.cs
// ScheduleGenerator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

class ScheduleEntry
{
    public string Subject { get; set; }
    public string Teacher { get; set; }
    public string Room { get; set; }
}

class ScheduleGenerator
{
    public int Days { get; set; }
    public int Periods { get; set; }
    public List<string> Subjects { get; set; }
    public List<string> Teachers { get; set; }
    public List<string> Rooms { get; set; }
    public DateTime StartTime { get; set; }
    public int Duration { get; set; }
    public Dictionary<int, List<ScheduleEntry>> Schedule { get; set; }

    public ScheduleGenerator(int days, int periods, List<string> subjects, List<string> teachers,
                            List<string> rooms, string startTime, int duration)
    {
        Days = days;
        Periods = periods;
        Subjects = subjects ?? new List<string> { "Math", "Physics", "Chemistry", "Biology", "History" };
        Teachers = teachers ?? new List<string> { "Smith", "Jones", "Williams", "Brown", "Davis" };
        Rooms = rooms ?? new List<string> { "101", "102", "103", "104", "105" };
        StartTime = DateTime.ParseExact(startTime, "HH:mm", null);
        Duration = duration;
        Schedule = new Dictionary<int, List<ScheduleEntry>>();
    }

    public void Generate()
    {
        if (Teachers.Count < Periods)
        {
            for (int i = Teachers.Count; i < Periods; i++)
                Teachers.Add($"T{i+1}");
        }
        if (Rooms.Count < Periods)
        {
            for (int i = Rooms.Count; i < Periods; i++)
                Rooms.Add($"R{i+1}");
        }
        Random rand = new Random();
        for (int day = 0; day < Days; day++)
        {
            var daySchedule = new List<ScheduleEntry>();
            var teachersShuffled = Teachers.OrderBy(x => rand.Next()).ToList();
            var roomsShuffled = Rooms.OrderBy(x => rand.Next()).ToList();
            for (int p = 0; p < Periods; p++)
            {
                string subject = Subjects[rand.Next(Subjects.Count)];
                daySchedule.Add(new ScheduleEntry
                {
                    Subject = subject,
                    Teacher = teachersShuffled[p % teachersShuffled.Count],
                    Room = roomsShuffled[p % roomsShuffled.Count]
                });
            }
            Schedule[day] = daySchedule;
        }
    }

    public void Display()
    {
        Console.WriteLine("\nGenerated Schedule:");
        Console.WriteLine(new string('-', 80));
        Console.WriteLine($"{"Day",-6} {"Period",-8} {"Time",-14} {"Subject",-12} {"Teacher",-12} Room");
        Console.WriteLine(new string('-', 80));
        for (int day = 0; day < Days; day++)
        {
            for (int p = 0; p < Periods; p++)
            {
                var entry = Schedule[day][p];
                var start = StartTime.AddMinutes(p * Duration);
                var end = start.AddMinutes(Duration);
                string timeStr = $"{start:HH:mm}-{end:HH:mm}";
                Console.WriteLine($"{day+1,-6} {p+1,-8} {timeStr,-14} {entry.Subject,-12} {entry.Teacher,-12} {entry.Room}");
            }
        }
    }

    public void ExportCSV(string filename)
    {
        using var writer = new StreamWriter(filename);
        writer.WriteLine("Day,Period,Time,Subject,Teacher,Room");
        for (int day = 0; day < Days; day++)
        {
            for (int p = 0; p < Periods; p++)
            {
                var entry = Schedule[day][p];
                var start = StartTime.AddMinutes(p * Duration);
                var end = start.AddMinutes(Duration);
                string timeStr = $"{start:HH:mm}-{end:HH:mm}";
                writer.WriteLine($"{day+1},{p+1},{timeStr},{entry.Subject},{entry.Teacher},{entry.Room}");
            }
        }
    }

    public void ExportJSON(string filename)
    {
        var data = new List<object>();
        for (int day = 0; day < Days; day++)
        {
            var periods = new List<object>();
            for (int p = 0; p < Periods; p++)
            {
                var entry = Schedule[day][p];
                var start = StartTime.AddMinutes(p * Duration);
                var end = start.AddMinutes(Duration);
                string timeStr = $"{start:HH:mm}-{end:HH:mm}";
                periods.Add(new { period = p+1, time = timeStr, subject = entry.Subject, teacher = entry.Teacher, room = entry.Room });
            }
            data.Add(new { day = day+1, periods });
        }
        string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
    }

    static void Main(string[] args)
    {
        if (args.Length > 0)
        {
            // CLI mode
            var dict = new Dictionary<string, string>();
            for (int i = 0; i < args.Length; i++)
            {
                if (args[i].StartsWith("--"))
                {
                    string key = args[i].Substring(2);
                    if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                        dict[key] = args[++i];
                    else
                        dict[key] = "true";
                }
            }
            int days = dict.ContainsKey("days") ? int.Parse(dict["days"]) : 5;
            int periods = dict.ContainsKey("periods") ? int.Parse(dict["periods"]) : 6;
            var subjects = dict.ContainsKey("subjects") ? dict["subjects"].Split(',').Select(s => s.Trim()).ToList() : null;
            var teachers = dict.ContainsKey("teachers") ? dict["teachers"].Split(',').Select(s => s.Trim()).ToList() : null;
            var rooms = dict.ContainsKey("rooms") ? dict["rooms"].Split(',').Select(s => s.Trim()).ToList() : null;
            string startTime = dict.ContainsKey("start") ? dict["start"] : "09:00";
            int duration = dict.ContainsKey("duration") ? int.Parse(dict["duration"]) : 45;
            string output = dict.ContainsKey("output") ? dict["output"] : null;
            var gen = new ScheduleGenerator(days, periods, subjects, teachers, rooms, startTime, duration);
            gen.Generate();
            gen.Display();
            if (output != null)
            {
                if (output.EndsWith(".json"))
                    gen.ExportJSON(output);
                else
                    gen.ExportCSV(output);
                Console.WriteLine($"Saved to {output}");
            }
        }
        else
        {
            Console.WriteLine("=== Schedule Generator ===");
            Console.Write("Enter number of days (1-7): ");
            int days = int.Parse(Console.ReadLine() ?? "5");
            Console.Write("Enter number of periods per day (1-8): ");
            int periods = int.Parse(Console.ReadLine() ?? "6");
            Console.Write("Enter subjects (comma-separated): ");
            var subjects = Console.ReadLine()?.Split(',').Select(s => s.Trim()).ToList() ?? new List<string>();
            if (subjects.Count == 0) subjects = new List<string> { "Math", "Physics", "Chemistry", "Biology", "History" };
            Console.Write("Enter teachers (comma-separated): ");
            var teachers = Console.ReadLine()?.Split(',').Select(s => s.Trim()).ToList() ?? new List<string>();
            if (teachers.Count == 0) teachers = new List<string> { "Smith", "Jones", "Williams", "Brown", "Davis" };
            Console.Write("Enter rooms (comma-separated): ");
            var rooms = Console.ReadLine()?.Split(',').Select(s => s.Trim()).ToList() ?? new List<string>();
            if (rooms.Count == 0) rooms = new List<string> { "101", "102", "103", "104", "105" };
            Console.Write("Start time (HH:MM, default 09:00): ");
            string startTime = Console.ReadLine();
            if (string.IsNullOrEmpty(startTime)) startTime = "09:00";
            Console.Write("Period duration (minutes, default 45): ");
            int duration = int.Parse(Console.ReadLine() ?? "45");
            var gen = new ScheduleGenerator(days, periods, subjects, teachers, rooms, startTime, duration);
            gen.Generate();
            gen.Display();
            Console.Write("Export to file? (y/n): ");
            if (Console.ReadLine()?.ToLower() == "y")
            {
                Console.Write("Filename (e.g., schedule.csv or schedule.json): ");
                string fname = Console.ReadLine();
                if (fname.EndsWith(".json"))
                    gen.ExportJSON(fname);
                else
                    gen.ExportCSV(fname);
                Console.WriteLine($"Saved to {fname}");
            }
        }
    }
}

File: ScheduleGenerator.java
// ScheduleGenerator.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import com.google.gson.*;

class ScheduleEntry {
    String subject;
    String teacher;
    String room;
}

class PeriodData {
    int period;
    String time;
    String subject;
    String teacher;
    String room;
}

class DayData {
    int day;
    List<PeriodData> periods;
}

public class ScheduleGenerator {
    private int days;
    private int periods;
    private List<String> subjects;
    private List<String> teachers;
    private List<String> rooms;
    private LocalTime startTime;
    private int duration;
    private Map<Integer, List<ScheduleEntry>> schedule = new HashMap<>();
    private Random rand = new Random();

    public ScheduleGenerator(int days, int periods, List<String> subjects, List<String> teachers,
                             List<String> rooms, String startTime, int duration) {
        this.days = days;
        this.periods = periods;
        this.subjects = subjects != null && !subjects.isEmpty() ? subjects :
                Arrays.asList("Math", "Physics", "Chemistry", "Biology", "History");
        this.teachers = teachers != null && !teachers.isEmpty() ? teachers :
                Arrays.asList("Smith", "Jones", "Williams", "Brown", "Davis");
        this.rooms = rooms != null && !rooms.isEmpty() ? rooms :
                Arrays.asList("101", "102", "103", "104", "105");
        this.startTime = LocalTime.parse(startTime, DateTimeFormatter.ofPattern("HH:mm"));
        this.duration = duration;
    }

    public void generate() {
        if (teachers.size() < periods) {
            for (int i = teachers.size(); i < periods; i++)
                teachers.add("T" + (i+1));
        }
        if (rooms.size() < periods) {
            for (int i = rooms.size(); i < periods; i++)
                rooms.add("R" + (i+1));
        }
        for (int day = 0; day < days; day++) {
            List<ScheduleEntry> daySchedule = new ArrayList<>();
            List<String> teachersShuffled = new ArrayList<>(teachers);
            Collections.shuffle(teachersShuffled, rand);
            List<String> roomsShuffled = new ArrayList<>(rooms);
            Collections.shuffle(roomsShuffled, rand);
            for (int p = 0; p < periods; p++) {
                String subject = subjects.get(rand.nextInt(subjects.size()));
                ScheduleEntry entry = new ScheduleEntry();
                entry.subject = subject;
                entry.teacher = teachersShuffled.get(p % teachersShuffled.size());
                entry.room = roomsShuffled.get(p % roomsShuffled.size());
                daySchedule.add(entry);
            }
            schedule.put(day, daySchedule);
        }
    }

    public void display() {
        System.out.println("\nGenerated Schedule:");
        System.out.println(new String(new char[80]).replace('\0', '-'));
        System.out.printf("%-6s %-8s %-14s %-12s %-12s %s%n", "Day", "Period", "Time", "Subject", "Teacher", "Room");
        System.out.println(new String(new char[80]).replace('\0', '-'));
        for (int day = 0; day < days; day++) {
            for (int p = 0; p < periods; p++) {
                ScheduleEntry entry = schedule.get(day).get(p);
                LocalTime start = startTime.plusMinutes(p * duration);
                LocalTime end = start.plusMinutes(duration);
                String timeStr = String.format("%02d:%02d-%02d:%02d", start.getHour(), start.getMinute(), end.getHour(), end.getMinute());
                System.out.printf("%-6d %-8d %-14s %-12s %-12s %s%n", day+1, p+1, timeStr, entry.subject, entry.teacher, entry.room);
            }
        }
    }

    public void exportCSV(String filename) throws IOException {
        try (PrintWriter pw = new PrintWriter(filename)) {
            pw.println("Day,Period,Time,Subject,Teacher,Room");
            for (int day = 0; day < days; day++) {
                for (int p = 0; p < periods; p++) {
                    ScheduleEntry entry = schedule.get(day).get(p);
                    LocalTime start = startTime.plusMinutes(p * duration);
                    LocalTime end = start.plusMinutes(duration);
                    String timeStr = String.format("%02d:%02d-%02d:%02d", start.getHour(), start.getMinute(), end.getHour(), end.getMinute());
                    pw.printf("%d,%d,%s,%s,%s,%s%n", day+1, p+1, timeStr, entry.subject, entry.teacher, entry.room);
                }
            }
        }
    }

    public void exportJSON(String filename) throws IOException {
        List<DayData> data = new ArrayList<>();
        for (int day = 0; day < days; day++) {
            DayData dd = new DayData();
            dd.day = day+1;
            dd.periods = new ArrayList<>();
            for (int p = 0; p < periods; p++) {
                ScheduleEntry entry = schedule.get(day).get(p);
                LocalTime start = startTime.plusMinutes(p * duration);
                LocalTime end = start.plusMinutes(duration);
                String timeStr = String.format("%02d:%02d-%02d:%02d", start.getHour(), start.getMinute(), end.getHour(), end.getMinute());
                PeriodData pd = new PeriodData();
                pd.period = p+1;
                pd.time = timeStr;
                pd.subject = entry.subject;
                pd.teacher = entry.teacher;
                pd.room = entry.room;
                dd.periods.add(pd);
            }
            data.add(dd);
        }
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(data);
        Files.write(Paths.get(filename), json.getBytes());
    }

    public static void main(String[] args) throws Exception {
        if (args.length > 0) {
            // CLI mode
            Map<String, String> params = new HashMap<>();
            for (int i = 0; i < args.length; i++) {
                if (args[i].startsWith("--")) {
                    String key = args[i].substring(2);
                    if (i+1 < args.length && !args[i+1].startsWith("--"))
                        params.put(key, args[++i]);
                    else
                        params.put(key, "true");
                }
            }
            int days = params.containsKey("days") ? Integer.parseInt(params.get("days")) : 5;
            int periods = params.containsKey("periods") ? Integer.parseInt(params.get("periods")) : 6;
            List<String> subjects = params.containsKey("subjects") ? Arrays.asList(params.get("subjects").split(",")) : null;
            List<String> teachers = params.containsKey("teachers") ? Arrays.asList(params.get("teachers").split(",")) : null;
            List<String> rooms = params.containsKey("rooms") ? Arrays.asList(params.get("rooms").split(",")) : null;
            String startTime = params.getOrDefault("start", "09:00");
            int duration = params.containsKey("duration") ? Integer.parseInt(params.get("duration")) : 45;
            String output = params.get("output");
            ScheduleGenerator gen = new ScheduleGenerator(days, periods, subjects, teachers, rooms, startTime, duration);
            gen.generate();
            gen.display();
            if (output != null) {
                if (output.endsWith(".json"))
                    gen.exportJSON(output);
                else
                    gen.exportCSV(output);
                System.out.println("Saved to " + output);
            }
        } else {
            // Interactive
            Scanner scanner = new Scanner(System.in);
            System.out.println("=== Schedule Generator ===");
            System.out.print("Enter number of days (1-7): ");
            int days = scanner.nextInt();
            System.out.print("Enter number of periods per day (1-8): ");
            int periods = scanner.nextInt();
            scanner.nextLine(); // consume newline
            System.out.print("Enter subjects (comma-separated): ");
            String subjectsLine = scanner.nextLine();
            List<String> subjects = subjectsLine.isEmpty() ? null : Arrays.asList(subjectsLine.split(","));
            System.out.print("Enter teachers (comma-separated): ");
            String teachersLine = scanner.nextLine();
            List<String> teachers = teachersLine.isEmpty() ? null : Arrays.asList(teachersLine.split(","));
            System.out.print("Enter rooms (comma-separated): ");
            String roomsLine = scanner.nextLine();
            List<String> rooms = roomsLine.isEmpty() ? null : Arrays.asList(roomsLine.split(","));
            System.out.print("Start time (HH:MM, default 09:00): ");
            String startTime = scanner.nextLine();
            if (startTime.isEmpty()) startTime = "09:00";
            System.out.print("Period duration (minutes, default 45): ");
            int duration = scanner.nextInt();
            scanner.nextLine();
            ScheduleGenerator gen = new ScheduleGenerator(days, periods, subjects, teachers, rooms, startTime, duration);
            gen.generate();
            gen.display();
            System.out.print("Export to file? (y/n): ");
            if (scanner.nextLine().toLowerCase().equals("y")) {
                System.out.print("Filename (e.g., schedule.csv or schedule.json): ");
                String fname = scanner.nextLine();
                if (fname.endsWith(".json"))
                    gen.exportJSON(fname);
                else
                    gen.exportCSV(fname);
                System.out.println("Saved to " + fname);
            }
            scanner.close();
        }
    }
}
